File: gareth-execution-ri/src/main/java/org/craftsmenlabs/gareth/execution/parser/ParsedDefinitionFactory.java
package org.craftsmenlabs.gareth.execution.parser;

import java.lang.annotation.Annotation;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.time.Duration;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Stream;
import org.craftsmenlabs.gareth.api.annotation.*;
import org.craftsmenlabs.gareth.api.exception.GarethDefinitionParseException;
import org.craftsmenlabs.gareth.api.exception.GarethExperimentParseException;
import org.craftsmenlabs.gareth.execution.invoker.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ParsedDefinitionFactory {

    private static Logger LOGGER = LoggerFactory.getLogger(ParsedDefinitionFactory.class);
    private final DefinitionFactory _definitionFactory;

    public ParsedDefinitionFactory(final DefinitionFactory definitionFactory) {
        this._definitionFactory = definitionFactory;
    }

    public ParsedDefinition parse(final Class clazz) throws GarethExperimentParseException {
        if (clazz == null) throw new IllegalArgumentException("Class cannot be null");
        final ParsedDefinition parsedDefinition = new ParsedDefinition();
        Stream.of(clazz.getMethods())
              .forEach(m -> parseMethod(m, parsedDefinition));
        return parsedDefinition;
    }

    /**
     * Parse a single method
     *
     * @param method
     * @param definition
     */
    private void parseMethod(final Method method, final ParsedDefinition definition) {
        Optional<Baseline> baseline = getAnnotation(method, Baseline.class);
        if (baseline.isPresent()) {
            definition.getBaselineDefinitions()
                      .put(baseline.get().glueLine(), createMethod(method, baseline.get().glueLine()));
        }
        Optional<Assume> assume = getAnnotation(method, Assume.class);
        if (assume.isPresent()) {
            definition.getAssumeDefinitions()
                      .put(assume.get().glueLine(), createMethod(method, assume.get().glueLine()));
        }
        Optional<Success> sucess = getAnnotation(method, Success.class);
        if (sucess.isPresent()) {
            definition.getSuccessDefinitions()
                      .put(sucess.get().glueLine(), createMethod(method, sucess.get().glueLine()));
        }
        Optional<Failure> failure = getAnnotation(method, Failure.class);
        if (failure.isPresent()) {
            definition.getFailureDefinitions()
                      .put(failure.get().glueLine(), createMethod(method, failure.get().glueLine()));
        }
        Optional<Time> time = getAnnotation(method, Time.class);
        if (time.isPresent()) {
            registerDuration(method, time.get().glueLine(), definition.getTimeDefinitions());
        }
    }

    public <T> Optional<T> getAnnotation(Method method, Class<T> annotationClass) {
        Optional<Annotation> first = Stream.of(method.getDeclaredAnnotations())
                                           .filter(an -> an.annotationType().getName()
                                                           .equals(annotationClass.getName())).findFirst();
        return (Optional<T>) first;
    }

    private RegexMethodDescriptor createMethod(final Method method, final String glueLine) {
        if (isValidMethod(method)) {
            LOGGER.info("Found valid method {} for glueline {}", method.getName(), glueLine);
            return new RegexMethodDescriptor(glueLine, method, 0, hasStorageParameter(method));

        } else {
            throw new IllegalStateException(String
                    .format("Method %s with glue line '%s' is not a valid method (no void return type)", method
                            .getName(), glueLine));
        }
    }

    /**
     * Register duration based on method outcome
     *
     * @param method
     * @param glueLine
     * @param durationMap
     */
    private void registerDuration(final Method method, final String glueLine, final Map<String, Duration> durationMap) throws GarethDefinitionParseException {
        if (isTimeMethod(method)) {
            try {
                final Object tmpDefinition = _definitionFactory.getInstanceForClass(method.getDeclaringClass());
                durationMap.put(glueLine, (Duration) method.invoke(tmpDefinition));
            } catch (final IllegalAccessException | InstantiationException | InvocationTargetException e) {
                throw new GarethDefinitionParseException(e);
            }
        } else {
            throw new IllegalStateException(String
                    .format("Method %s with glue line '%s' is not a valid method (no duration return type)", method
                            .getName(), glueLine));
        }
    }

    private boolean isValidMethod(final Method method) {
        return (method.getReturnType().equals(Void.class)
                || method.getReturnType().equals(Void.TYPE));
    }

    private boolean hasStorageParameter(Method method) {
        return method.getParameterCount() > 0 && method.getParameters()[0]
                .getParameterizedType() == DefaultStorage.class;
    }

    private boolean isTimeMethod(final Method method) {
        return method.getReturnType().isAssignableFrom(Duration.class);
    }

    private boolean isRegexTimeMethod(final Method method) {
        return isTimeMethod(method) &&
                method.getParameterCount() == 2 &&
                method.getParameterTypes()
                      .equals(new Object[]{Integer.class, String.class});
    }
}

File: gareth-execution-ri/src/main/kotlin/org/craftsmenlabs/gareth/execution/definitions/ParsedDefinitionFactory.kt
package org.craftsmenlabs.gareth.execution.definitions

import org.craftsmenlabs.gareth.validator.*
import org.slf4j.LoggerFactory
import java.lang.reflect.Method
import java.time.Duration

class ParsedDefinitionFactory {

    private val log = LoggerFactory.getLogger(ParsedDefinitionFactory::class.java)

    fun parse(clazz: Class<*>): ParsedDefinition {
        val parsedDefinition = ParsedDefinition()
        clazz.methods.forEach { m -> parseMethod(m, parsedDefinition) }
        return parsedDefinition
    }

    /**
     * Parse a single method

     * @param method
     * *
     * @param definition
     */
    private fun parseMethod(method: Method, definition: ParsedDefinition) {
        val baseline = getAnnotation(method, Baseline::class.java)
        if (baseline != null) {
            definition.baselineDefinitions.put(baseline.glueLine, createMethod(method, baseline.glueLine, baseline.description, baseline.humanReadable))
        }
        val assume = getAnnotation(method, Assume::class.java)
        if (assume != null) {
            definition.assumeDefinitions.put(assume.glueLine, createMethod(method, assume.glueLine, assume.description, assume.humanReadable, true))
        }
        val sucess = getAnnotation(method, Success::class.java)
        if (sucess != null) {
            definition.successDefinitions.put(sucess.glueLine, createMethod(method, sucess.glueLine, sucess.description, sucess.humanReadable))
        }
        val failure = getAnnotation(method, Failure::class.java)
        if (failure != null) {
            definition.failureDefinitions.put(failure.glueLine, createMethod(method, failure.glueLine, failure.description, failure.humanReadable))
        }
        val time = getAnnotation(method, Time::class.java)
        if (time != null) {
            definition.timeDefinitions.put(time.glueLine, createMethod(method,time.glueLine,time.description,time.humanReadable))
        }
    }

    fun <T> getAnnotation(method: Method, annotationClass: Class<T>): T? {
        return method.declaredAnnotations.filter { it.annotationClass.simpleName.equals(annotationClass.simpleName) }.firstOrNull() as T
    }

    private fun createMethod(method: Method,
                             glueLine: String,
                             description: String? = null,
                             humanReadable: String? = null,
                             expectBoolean: Boolean = false): InvokableMethod {
        val isBoolean = method.returnType == Boolean::class.java
        val isVoid = method.returnType != Void::class.java || method.returnType != Void.TYPE
        if (expectBoolean && !isBoolean)
            throw GarethIllegalDefinitionException("Method return type must be boolean but is ${method.returnType}")
        if (!expectBoolean && !isVoid)
            throw GarethIllegalDefinitionException("Method return type must be void but is ${method.returnType}")
        fun nonEmpty(str: String?): String? = if (str == null || str.isBlank()) null else str!!
        return InvokableMethod(
                glueLine = glueLine,
                description = description,
                method = method,
                humanReadable = nonEmpty(humanReadable) ?: nonEmpty(description) ?: glueLine,
                runcontextParameter = hasRunContextParameter(method))
    }

    /**
     * Register duration based on method outcome

     * @param method
     * *
     * @param glueLine
     * *
     * @param durationMap
     */


    private fun hasRunContextParameter(method: Method): Boolean {
        return method.parameterCount > 0 && InvokableMethod.isContextParameter(method.parameterTypes[0])// == "org.craftsmenlabs.gareth.validator.model.RunContext"
    }

    private fun isTimeMethod(method: Method): Boolean {
        return method.returnType.isAssignableFrom(Duration::class.java)
    }

}